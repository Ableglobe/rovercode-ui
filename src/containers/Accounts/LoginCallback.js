import React, { Component, Fragment } from 'react';
import { Redirect } from 'react-router';
import { Loader } from 'semantic-ui-react';
import { connect } from 'react-redux';
import axios from 'axios';
import queryString from 'query-string';
import PropTypes from 'prop-types';
import { withCookies, Cookies } from 'react-cookie';
import jwtDecode from 'jwt-decode';

import { updateValidAuth as actionUpdateValidAuth } from '@/actions/auth';
import { updateUser as actionUpdateUser } from '@/actions/user';

const mapDispatchToProps = dispatch => ({
  updateUser: data => dispatch(actionUpdateUser({ ...data, isSocial: true })),
  updateValidAuth: isValidAuth => dispatch(actionUpdateValidAuth(isValidAuth)),
});

class LoginCallback extends Component {
  constructor(props) {
    super(props);

    this.state = {
      loading: true,
      loginSuccess: false,
    };
  }

  componentDidMount() {
    const {
      cookies, location, match, updateUser, updateValidAuth,
    } = this.props;
    const queryParams = queryString.parse(location.search);

    return axios.post(`/jwt/auth/social/${match.params.service}/login/`, {
      code: queryParams.code,
      state: queryParams.state,
    })
      .then((response) => {
        updateUser(jwtDecode(response.data.token));
        cookies.set('auth_jwt', response.data.token, { path: '/' });
        updateValidAuth(true);
        this.setState({
          loading: false,
          loginSuccess: true,
        });
      })
      .catch(() => {
        this.setState({
          loading: false,
          loginSuccess: false,
        });
      });
  }

  redirect = () => {
    const { loading, loginSuccess } = this.state;

    if (loading) {
      return <Loader content="Logging in..." />;
    }

    if (loginSuccess) {
      return <Redirect to="/" />;
    }

    return <Redirect to="/accounts/login" />;
  }

  render() {
    return (
      <Fragment>
        { this.redirect() }
      </Fragment>
    );
  }
}

LoginCallback.propTypes = {
  cookies: PropTypes.instanceOf(Cookies).isRequired,
  location: PropTypes.shape({
    search: PropTypes.string.isRequired,
  }).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      service: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  updateUser: PropTypes.func.isRequired,
  updateValidAuth: PropTypes.func.isRequired,
};

export default withCookies(connect(null, mapDispatchToProps)(LoginCallback));
