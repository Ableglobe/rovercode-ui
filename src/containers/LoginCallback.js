import React, { Component, Fragment } from 'react';
import { Redirect } from 'react-router';
import { Loader } from 'semantic-ui-react';
import axios from 'axios';
import queryString from 'query-string';
import PropTypes from 'prop-types';
import { withCookies, Cookies } from 'react-cookie';

class LoginCallback extends Component {
  constructor(props) {
    super(props);

    this.state = {
      loading: true,
      loginSuccess: false,
    };
  }

  componentDidMount() {
    const { cookies, location, match } = this.props;
    const queryParams = queryString.parse(location.search);

    return axios.post(`/jwt/auth/social/${match.params.service}/login/`, {
      code: queryParams.code,
      state: queryParams.state,
    })
      .then((response) => {
        cookies.set('auth_jwt', response.data.token, { path: '/' });
        this.setState({
          loading: false,
          loginSuccess: true,
        });
      })
      .catch(() => {
        this.setState({
          loading: false,
          loginSuccess: false,
        });
      });
  }

  redirect = () => {
    const { loading, loginSuccess } = this.state;

    if (loading) {
      return <Loader content="Logging in..." />;
    }

    if (loginSuccess) {
      return <Redirect to="/" />;
    }

    return <Redirect to="/login" />;
  }

  render() {
    return (
      <Fragment>
        { this.redirect() }
      </Fragment>
    );
  }
}

LoginCallback.propTypes = {
  cookies: PropTypes.instanceOf(Cookies).isRequired,
  location: PropTypes.shape({
    search: PropTypes.string.isRequired,
  }).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      service: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
};

export default withCookies(LoginCallback);
