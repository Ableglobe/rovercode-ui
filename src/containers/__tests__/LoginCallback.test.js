import React from 'react';
import { mount } from 'enzyme';
import {
  MemoryRouter, Route, Switch,
} from 'react-router';
import { Loader } from 'semantic-ui-react';
import { CookiesProvider } from 'react-cookie';
import toJson from 'enzyme-to-json';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import LoginCallback from '../LoginCallback';

const mock = new MockAdapter(axios);

const initialEntry = {
  pathname: '/login/google/callback',
  search: '?state=1234&code=5678',
  key: '123456',
};

test('LoginCallback renders on the page with no errors', () => {
  const wrapper = mount(
    <CookiesProvider>
      <MemoryRouter initialEntries={[initialEntry]}>
        <Switch>
          <Route path="/login/:service/callback" component={LoginCallback} />
        </Switch>
      </MemoryRouter>
    </CookiesProvider>,
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});

test('LoginCallback displays loader while loading', () => {
  mock.reset();
  mock.onPost('/api/auth/social/google/login/').timeout();
  const wrapper = mount(
    <CookiesProvider>
      <MemoryRouter initialEntries={[initialEntry]}>
        <Switch>
          <Route path="/login/:service/callback" component={LoginCallback} />
        </Switch>
      </MemoryRouter>
    </CookiesProvider>,
  );

  wrapper.setState({
    loading: true,
  });
  wrapper.update();

  expect(wrapper.find(Loader).exists()).toBe(true);
});
