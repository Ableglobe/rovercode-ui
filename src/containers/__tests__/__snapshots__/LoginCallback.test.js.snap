// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginCallback renders on the page with no errors 1`] = `
<CookiesProvider>
  <MemoryRouter
    initialEntries={
      Array [
        Object {
          "key": "123456",
          "pathname": "/login/google/callback",
          "search": "?state=1234&code=5678",
        },
      ]
    }
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "123456",
              "pathname": "/login/google/callback",
              "search": "?state=1234&code=5678",
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "123456",
            "pathname": "/login/google/callback",
            "search": "?state=1234&code=5678",
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Switch>
        <Route
          component={[Function]}
          computedMatch={
            Object {
              "isExact": true,
              "params": Object {
                "service": "google",
              },
              "path": "/login/:service/callback",
              "url": "/login/google/callback",
            }
          }
          location={
            Object {
              "hash": "",
              "key": "123456",
              "pathname": "/login/google/callback",
              "search": "?state=1234&code=5678",
            }
          }
          path="/login/:service/callback"
        >
          <withCookies(Component)
            history={
              Object {
                "action": "POP",
                "block": [Function],
                "canGo": [Function],
                "createHref": [Function],
                "entries": Array [
                  Object {
                    "hash": "",
                    "key": "123456",
                    "pathname": "/login/google/callback",
                    "search": "?state=1234&code=5678",
                  },
                ],
                "go": [Function],
                "goBack": [Function],
                "goForward": [Function],
                "index": 0,
                "length": 1,
                "listen": [Function],
                "location": Object {
                  "hash": "",
                  "key": "123456",
                  "pathname": "/login/google/callback",
                  "search": "?state=1234&code=5678",
                },
                "push": [Function],
                "replace": [Function],
              }
            }
            location={
              Object {
                "hash": "",
                "key": "123456",
                "pathname": "/login/google/callback",
                "search": "?state=1234&code=5678",
              }
            }
            match={
              Object {
                "isExact": true,
                "params": Object {
                  "service": "google",
                },
                "path": "/login/:service/callback",
                "url": "/login/google/callback",
              }
            }
          >
            <LoginCallback
              allCookies={Object {}}
              cookies={
                Cookies {
                  "HAS_DOCUMENT_COOKIE": true,
                  "changeListeners": Array [
                    [Function],
                  ],
                  "cookies": Object {},
                  "hooks": undefined,
                }
              }
              history={
                Object {
                  "action": "POP",
                  "block": [Function],
                  "canGo": [Function],
                  "createHref": [Function],
                  "entries": Array [
                    Object {
                      "hash": "",
                      "key": "123456",
                      "pathname": "/login/google/callback",
                      "search": "?state=1234&code=5678",
                    },
                  ],
                  "go": [Function],
                  "goBack": [Function],
                  "goForward": [Function],
                  "index": 0,
                  "length": 1,
                  "listen": [Function],
                  "location": Object {
                    "hash": "",
                    "key": "123456",
                    "pathname": "/login/google/callback",
                    "search": "?state=1234&code=5678",
                  },
                  "push": [Function],
                  "replace": [Function],
                }
              }
              location={
                Object {
                  "hash": "",
                  "key": "123456",
                  "pathname": "/login/google/callback",
                  "search": "?state=1234&code=5678",
                }
              }
              match={
                Object {
                  "isExact": true,
                  "params": Object {
                    "service": "google",
                  },
                  "path": "/login/:service/callback",
                  "url": "/login/google/callback",
                }
              }
            >
              <Loader
                content="Logging in..."
              >
                <div
                  className="ui text loader"
                >
                  Logging in...
                </div>
              </Loader>
            </LoginCallback>
          </withCookies(Component)>
        </Route>
      </Switch>
    </Router>
  </MemoryRouter>
</CookiesProvider>
`;
